### PR TITLE
Feature book controller

### PR DESCRIPTION
**Pull request by JohnyTea**

### CODE REVIEW

File: Library.API/Controllers/BooksController.cs [hunk 1]
```diff
@@ -0,0 +1,97 @@
+﻿using Library.API.Data;
+using Library.API.Exceptions;
+namespace Library.API.Controllers;
+
+[Route("api/[controller]")]
+[ApiController]
+public class BookControler : ControllerBase
```
Comment (JCzekiel) [inline Library.API/Controllers/BooksController.cs:7]: nie moge

File: Library.API/Controllers/BooksController.cs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+{
+    private readonly ApplicationDataBaseContext _context;
```
Comment (JCzekiel) [inline Library.API/Controllers/BooksController.cs]: Man pls do not use fking "_" as private annotation...

File: Library.API/Controllers/BooksController.cs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    private readonly ILibraryService _library;
+
+    public BookControler(ApplicationDataBaseContext context)
+    {
+        _context = context;
+        _library = new LibraryService(_context);
+    }
+
+    // GET: api/Books
```
Comment (JCzekiel) [inline Library.API/Controllers/BooksController.cs]: Remove comment

File: Library.API/Controllers/BooksController.cs [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    [HttpGet]
+    public async Task<ActionResult<IEnumerable<Book>>> Get()
+    {
+        return await _library.Books.GetAllAsync();
+    }
+
+    // GET: api/Books/5
```
Comment (JCzekiel) [inline Library.API/Controllers/BooksController.cs]: Remove comment

File: Library.API/Controllers/BooksController.cs [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    [HttpGet("{id}")]
+    public async Task<ActionResult<Book>> Get(int id)
+    {
+        return await _library.Books.GetByIdAsync(id);
+    }
+
+    // POST: api/Users
```
Comment (JCzekiel) [inline Library.API/Controllers/BooksController.cs]: Remove comment

File: Library.API/Controllers/BooksController.cs [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+            Year = book.Year,
+            ISBN = book.ISBN,
+        };
+
+        await _library.Books.AddAsync(newBook);
+        await _context.SaveChangesAsync();
+        return CreatedAtAction("Get", new { id = newBook.Id }, newBook);
+    }
+
+    // PUT: api/Users/5
```
Comment (JCzekiel) [inline Library.API/Controllers/BooksController.cs]: Remove comment

File: Library.API/Controllers/BooksController.cs [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+        Book updatedBook = new()
+        {
+            Author = book.Author,
+            Title = book.Title,
+            Publisher = book.Publisher,
+            Year = book.Year,
+            ISBN = book.ISBN,
+        };
+
+        if (await BookExistAsync(id))
```
Comment (JCzekiel) [inline Library.API/Controllers/BooksController.cs:60]: Is bookExistAsync method really needed?

File: Library.API/Controllers/BooksController.cs [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        else
+        {
+            await _library.Books.AddAsync(updatedBook);
+        }
+
+        await _context.SaveChangesAsync();
+        return Ok();
+    }
+
+    // DELETE: api/Users/5
```
Comment (JCzekiel) [inline Library.API/Controllers/BooksController.cs]: Remove comment

File: Library.API/Controllers/UsersController.cs [hunk 9]
```diff
@@ -0,0 +1,111 @@
+﻿using Library.API.Data;
+
+namespace Library.API.Controllers;
+
+[Route("api/[controller]")]
+[ApiController]
+public class UsersController : ControllerBase
+{
+    private readonly ApplicationDataBaseContext _context;
```
Comment (JCzekiel) [inline Library.API/Controllers/UsersController.cs]: "_" :'(

File: Library.API/Controllers/UsersController.cs [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+            UserName = user.UserName,
+            BirthDate = user.BirthDate
+        };
+
+        await _library.Users.AddAsync(newUser);
+        await _context.SaveChangesAsync();
+        return CreatedAtAction("GetUser", new { id = newUser.Id }, newUser);
+    }
+
+    // PUT: api/Users/5
```
Comment (JCzekiel) [inline Library.API/Controllers/UsersController.cs]: Remove comment

File: Library.API/Data/BookDto.cs [hunk 11]
```diff
@@ -0,0 +1,20 @@
+﻿using System.ComponentModel.DataAnnotations;
+
+namespace Library.API.Data;
+
+public struct BookDto
+{
+    public string Author { get; set; }
```
Comment (JCzekiel) [inline Library.API/Data/BookDto.cs:7]: Can a book have no author :D ?

File: Library.API/Data/BorrowBookDto.cs [hunk 12]
```diff
@@ -0,0 +1,12 @@
+﻿using System.ComponentModel.DataAnnotations;
+
+namespace Library.API.Data;
+
+public class BorrowBookDto
```
Comment (JCzekiel) [inline Library.API/Data/BorrowBookDto.cs]: I believe, that it should be BorrowedBook

File: Library.API/Models/Book.cs [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+    public string? Publisher { get; set; }
+
+    public int Year { get; set; }
+
+    [Required]
+    [MinLength(13)]
+    [MaxLength(13)]
+    public string ISBN { get; set; }
+
+    public int? BorrowerID { get; set; }
```
Comment (JCzekiel) [inline Library.API/Models/Book.cs:26]: :(

File: Library.API/Services/BookService.cs [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+
+namespace Library.API.Services;
+
+public class BookService : IBookService
+{
+    private readonly ApplicationDataBaseContext _context;
+
+    public BookService(ApplicationDataBaseContext context)
+    {
+        _context = context;
```
Comment (JCzekiel) [inline Library.API/Services/BookService.cs:12]: "_" :'(

File: Library.API/Services/BookService.cs [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    public async Task<List<Book>> GetAllAsync()
+    {
+        var books = await _context.Books.Include(b => b.Borrower).ToListAsync();
+        if (books is null) {
```
Comment (JCzekiel) [inline Library.API/Services/BookService.cs:18]: I am sad

File: Library.API/Services/ILibraryService.cs [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+        public Task Return(int bookID);
+    }
+
+    public interface IUserService
+    {
+        public Task<List<User>> GetAllAsync();
+        public Task<User> GetByIdAsync(int id);
+        public Task AddAsync(User newUser);
+        public Task EditAsync(int id, User updatedUser);
+        public Task DeleteAsync(int id);
```
Comment (JCzekiel) [inline Library.API/Services/ILibraryService.cs]: ById blin

**Review by JCzekiel — COMMENTED** [review]
Hehe nice try

File: Library.API/Data/BookDto.cs (hunk 11 above)
Comment (JohnyTea) [reply Library.API/Data/BookDto.cs:7]: Yes, author can be unknown

**Pull request merged**